File: src/shortTerm.ts
// Short-term memory (STM): a bounded conversation buffer that mirrors an LLM's
// limited context window. When it overflows, the oldest half is summarized into a
// single summary message — the same "memory pressure" idea as MemGPT.

import type { ChatMessage } from "./types.ts";
import type { LLM } from "./llm.ts";

// Rough token estimate: ~4 characters per token
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4);
}

function messageTokens(m: ChatMessage): number {
  return estimateTokens(m.content) + 4; // +4 for role/formatting overhead
}

export class ConversationBuffer {
  private messages: ChatMessage[] = [];
  readonly maxTokens: number;

  constructor(maxTokens = 4000) {
    this.maxTokens = maxTokens;
  }

  add(message: ChatMessage): void {
    this.messages.push(message);
  }

  // All messages currently held (the working context)
  getMessages(): ChatMessage[] {
    return this.messages;
  }

  tokenCount(): number {
    return this.messages.reduce((sum, m) => sum + messageTokens(m), 0);
  }

  // Percentage of the token budget currently used 
  pressure(): number {
    return Math.round((this.tokenCount() / this.maxTokens) * 100);
  }

  isOverCap(): boolean {
    return this.tokenCount() > this.maxTokens;
  }

  /**
   * If over the token cap, summarize the oldest half of the (non-summary) messages
   * into a single system summary, replacing the originals. Returns true if it summarized.
   */
  async summarizeIfNeeded(llm: LLM): Promise<boolean> {
    if (!this.isOverCap() || this.messages.length < 4) return false;

    const half = Math.floor(this.messages.length / 2);
    const toSummarize = this.messages.slice(0, half);
    const remaining = this.messages.slice(half);

    const transcript = toSummarize
      .map((m) => `${m.role}: ${m.content}`)
      .join("\n");

    const { content } = await llm.chat([
      {
        role: "system",
        content:
          "Summarize the following conversation excerpt into a few compact bullet points, " +
          "preserving any facts, names, preferences, or decisions. Be terse.",
      },
      { role: "user", content: transcript },
    ]);

    const summaryMessage: ChatMessage = {
      role: "system",
      content: `[Conversation summary]\n${content}`,
    };
    this.messages = [summaryMessage, ...remaining];
    return true;
  }
}

File: src/short_term.py
from typing import Callable

from src.config import CHAT_MODEL, client

CHARS_PER_TOKEN = 4  # rough OpenAI rule-of-thumb for English text; no tokenizer library in requirements.txt


def estimate_tokens(text: str) -> int:
    return len(text) // CHARS_PER_TOKEN


def summarize_messages(messages: list[dict]) -> str:
    transcript = "\n".join(f"{m['role']}: {m['content']}" for m in messages)
    response = client.chat.completions.create(
        model=CHAT_MODEL,
        messages=[
            {
                "role": "system",
                "content": (
                    "Summarize the following conversation excerpt concisely, preserving any "
                    "facts, names, or preferences mentioned."
                ),
            },
            {"role": "user", "content": transcript},
        ],
    )
    return response.choices[0].message.content


class ConversationBuffer:
    def __init__(self, max_tokens: int = 4000, summarizer: Callable[[list[dict]], str] = summarize_messages):
        self.max_tokens = max_tokens
        self.messages: list[dict] = []
        self._summarizer = summarizer
        self.last_summary: str | None = None

    def add(self, role: str, content: str) -> bool:
        """Appends a message. Returns True if adding it triggered an overflow summarization."""
        self.messages.append({"role": role, "content": content})
        if self.token_count() > self.max_tokens:
            return self._summarize_oldest_half()
        return False

    def token_count(self) -> int:
        return sum(estimate_tokens(m["content"]) for m in self.messages)

    def usage_ratio(self) -> float:
        return self.token_count() / self.max_tokens if self.max_tokens else 0.0

    def _summarize_oldest_half(self) -> bool:
        split = len(self.messages) // 2
        if split == 0:
            return False
        oldest, rest = self.messages[:split], self.messages[split:]
        summary = self._summarizer(oldest)
        self.last_summary = summary
        self.messages = [{"role": "system", "content": f"[Conversation summary] {summary}"}] + rest
        return True
